### ISSUE TITLE
Drop std::auto_ptr usage

### ISSUE THREAD

**Issue by Kicer86** [Exiv2/exiv2#46]
std::auto_ptr is old, deprecated and I causes build problems on some platforms (VisualStudio with c++17 enabled has no std::auto_ptr).

I would suggest to replace it with std::unque_ptr (at least on some cmake switch).

**Comment by clanmills** [comment]
I agree totally.  We're planning a dot release Exiv2 v0.26.1 to include support for C++11 and will address your concern.  I'm closing your issue because this subject is "work-in-progress".

**Comment by Kicer86** [comment]
Thank you for information.
As I needed it now, I've made a fork and I've made proper substitutions.
If you would like to merge it or review it, let me know, as I had some concerns in a few code locations which you may be interested in.

**Comment by clanmills** [comment]
Thank You, Michał

I would like to review your suggestions.  Ben, who is a member of Team Exiv2, has code in a branch which I am intending to integrate to master.  However, I will be happy to look at your proposal to deal with this.  I'm reopening this issue to assure you of my attention!

**Comment by Kicer86** [comment]
OK :)

When I was replacing auto_ptr I think I've found a bug - please take a look at TiffComponent* TiffDirectory::doAddPath(uint16_t tag, TiffPath& tiffPath, TiffComponent* const pRoot, TiffComponent::AutoPtr object) function.  'object' (which is an auto_ptr) is being copied twice - line 680 and then 698. Is this intentional? The same for line 733 and 805.

**Comment by clanmills** [comment]
I didn't write that code.  It sounds like a bug.  The tiffmumble.cpp code is ingenious and difficult to understand.  However, it has stood the test of time.  It has been there for years and almost no bugs have been reported which required changes.

```
511 rmills@rmillsmbp:~/gnu/exiv2/0.26 $ svn blame src/tiffcomposite.cpp | grep -v -e huggel
  3479 robinwmill #include "config.h"
  4407       tbeu         : TiffEntryBase(tag, group),
  4407       tbeu         elByteOrder_(invalidByteOrder)
  4407       tbeu         elDef_.idx_ = 0;
  4407       tbeu         elDef_.tiffType_ = ttUndefined;
  4407       tbeu         elDef_.count_ = 0;
  2962 robinwmill         // http://dev.exiv2.org/boards/3/topics/1337 change unsignedByte to signedByte
  2962 robinwmill         // Exif.NikonAFT.AFFineTuneAdj || Exif.Pentax.Temperature
  4407       tbeu         if ( ti == Exiv2::unsignedByte ) {
  4407       tbeu             if ( (tag == 0x0002 && group == nikonAFTId ) || (tag == 0x0047 && group == pentaxId) ) {
  4407       tbeu                 ti = Exiv2::signedByte;
  4407       tbeu             }
  2960 robinwmill         }
512 rmills@rmillsmbp:~/gnu/exiv2/0.26 $ 
```

**Comment by Kicer86** [comment]
Frankly speaking I'm observing random crashes when I use exiv2 on tiff files (in multithread program). Now I've Exiv2 builded with debug information so I'll try to hunt it down.
I may also try to figure out what's going on with this double copy of auto_ptr (if I undersand it :) )

**Comment by clanmills** [comment]
I am not aware of multi-threading tiff crashes, so I will interested to hear what you discover.  I believe the code is robust because very few bugs have been reported against that code.  I know that digiKam is a multi-threaded application that reads images including tiffs.  

The tiff parsing code is the part of Exiv2 which I understand least.  It's based on the Visitor Pattern described in Addison & Wesley's boot "Design Patterns"  https://books.google.co.uk/books/about/Design_Patterns.html?id=6oHuKQe3TjQC&printsec=frontcover&source=kp_read_button&redir_esc=y#v=onepage&q&f=false

Tiff structures are used extensively to store metadata.  Even in a JPEG, there is a embedded tiff structure which contains the Exif metadata.  Tiff structures can be recursive.  You will files which have tiff structures embedded in the embedded tiff structure.  I believe the strategy is to define call backs which are invoked by the parser when selected tags are encountered.

I'm happy to help you with this task.  If you feel I can be off assistance, please contact me and I'll try to help.

**Comment by Kicer86** [comment]
I've figured out what's causing crash in multithread environment: tagName() from image.cpp used by Image::printIFDStructure is not thread safe. It's due to statics used inside. In my case when two or more threads enter this function simultanously they try to initialize map which causes crash.

Fixing it should be simple and I can do it. I'd move tagName into Image class and I'd make all these static variables also Image's members. It will fix threadsafe problems if user doesn't use the same Image object from many threads but uses one instance per thread as in my case. This is simplest solution, yet will definitely have impact on performance (is it relevant?). Another thing is that it would also 'break' current code structure (similar functions like typeName or typeValid would stay global). 

What do you think about it?

**Comment by clanmills** [comment]
Thanks for fixing this.  My apologies.  That code was initially a debugging function to be used by the Exiv2 command line program.  Then I discovered a totally different purpose for it which I will explain in a moment.  In the heat of battle I neglected to move the statics into the class as you have suggested.  Sure, please fix that and accept my apologies for this.

The new purpose for which that code is used is to read the metadata into memory.  We normally use memory mapping to read tiff files.  The tiff visitor code operates on memory addresses.  For internet files, we want to avoid copying the whole file into contiguous memory.  So we allocate a block of memory which is the same size as the file.  Then we use the function to read the meta using file seeks.  This ensure that the tiff visitor will encounter the metadata without copying the whole
Files over the network.

This is a really great find.  Thank you very much for working on this.  I feel guilty about not removing the statics.  That's not good!

Robin
http://clanmills.com
Not sent from my iPhone

**Comment by Kicer86** [comment]
Hi Robin.

There's no need for you to apologize, I'm glad I could help improving your project :)
I can share my fix this evening as a pull request.

Anyway nice trick with seeking, I'll take a look at this solution, it may be useful for me one day :)

Regards.

**Comment by Kicer86** [comment]
Here it goes #62

**Comment by Kicer86** [comment]
Optimizing code is always fun, yet it's very easy to fall into some sophisticated traps :)

I'm glad I could help, and I would like to do some more :)

Regards
Michał

**Comment by clanmills** [comment]
I would be delighted to have your help.  Luis, Ben and I are working on a “dot” release v0.26.1.  Leave us to deal with that.

Have a look at the proposed projects for v0.27 and see if there’s something that interests you.  And you’re welcome to suggest projects which are not on the list.
http://dev.exiv2.org/news/3 <http://dev.exiv2.org/news/3>

The Linux security people dropped a little bag of goodies on us this week (about 12 security issues).  I knew that was coming.  You’ve already fixed several of them.  Mostly horrible stuff!

Robin

**Comment by D4N** [comment]
Just as a note: [clang-tidy](https://clang.llvm.org/extra/clang-tidy/index.html) with the modernize checks could come in handy when moving to C++11.

**Comment by D4N** [comment]
Afaik we have deferred C++11 support to 0.28, so maybe we should think about either moving to C++11 earlier or supplying our own version of `std::unique_ptr<T>` or `std::shared_ptr<T>`. Both are not optimal, so we'll have to choose the lesser evil.

**Comment by clanmills** [comment]
Perhaps we should move to C++11 in v0.27.  In the past, I've announced technology to be "deprecated" in advance.  However, you and Luis have introduced stuff that really challenges that conservative approach.  For example safe_op.hpp includes Intsafe.h which breaks VS2005 and other elderly versions of Visual Studio.

Back in summer 2017, my intention was to have a v0.26.1 release that would 'patch' v0.26 and I was intending to deprecate C++98 for v0.27.  The security situation has changed everything.  You and Luis (and others) have done great work to deal with that and I'm happy to defer to your judgement about how to advance from here.

So, how about we decide to move to C++11 with v0.27.  If somebody really wants v0.27 to build using C++98, they can undertake the effort.

**Comment by D4N** [comment]
I have tried to stay on C++03 as much as I can. Concerning the issue
with VS2005 and safe_op.hpp, we can remove the include of Intsafe.h for
old versions of Visual Studio (the code supports that easily). Could you
tell me to what value VS2005 defines the Macro _MSC_VER (I think it
should be 500)? Then I'll add an include switch.

Since we are mostly "serving" programmers, we could try asking our
users? When most of them prefer C++ > 11 we could decide to move forward
earlier or we could defer the move further if many of them don't want to
move to modern C++ yet. But beside the KDE project, I don't know who
uses exiv2.

But I'd like to add one thing: once we move to C++11, there is no going
back. There are a lot of features that are impossible to emulate in
C++03/98 (like static_assert of values obtained from constexpr
functions, rvalue references and variadic templates to name a few). Once
we start using these features or features from the new standard library
(like shared_ptr, unique_ptr, etc), users of exiv2 must start using C++11.

But then again, C++11 or even C++14 is the new standard in gcc/clang and
afaik also in MSVC (I think I have heard somewhere that you can't even
go back to C++03 in the newest MSVC, but I find that hard to
believe). So the change is in some ways inevitable.

**Comment by clanmills** [comment]
I think we're moving forward to C++11 and was intended for v0.27 in 2018.  The plans for v0.26.1 died with the invasion of the security people.  I rebranded "0.26.1" as 0.27 and "0.27" as 0.28.  The time has come for this change.

I can discuss this with Exiv2 users on https://discuss.pixls.us  An Exiv2 status announcement is due.  I was hoping to introduce you and Luis and Michal in that announcement and to say that we're going to have an "Exiv2 Developer's Meeting" in England this year.  Thomas and Michal are interested in coming and I'm wondering about you and Luis.

If folks want to use C++98, they can continue to use v0.26.  Thanks to great work from you and the community, security patches are available for v0.26 and v0.25.

The values of _MSC_VER for different versions of Visual Studio are in include/exiv2/config.h

**Comment by piponazo** [comment]
This was finally implemented in #601
